wdl-lint/src/rules/nonmatching_output.rs

```diff
@@ -0,0 +1,249 @@
+    context: TaskOrWorkflow,
+) {
+    // Get the `outputs` entries from the `meta` section.
+    if let Some(meta_outputs_key) = meta
+        .items()
+        .find(|entry| entry.name().as_str() == "outputs")
+    {
+        let actual: IndexMap<_, _> = meta_outputs_key
+            .value()
+            .unwrap_object()
+            .items()
+            .map(|entry| {
+                (
+                    entry.name().syntax().to_string(),
+                    entry.syntax().text_range().to_span(),
+                )
+            })
+            .collect();
+
+        // Get the declared outputs.
+        let expected: IndexMap<_, _> = outputs
+            .declarations()
+            .map(|d| {
+                (
+                    d.name().syntax().to_string(),
+                    d.syntax().text_range().to_span(),
+                )
+            })
+            .collect();
+
+        let mut extra_found = false;
+
+        // Check for entries missing from `meta`.
+        for (name, span) in &expected {
+            if !actual.contains_key(name) {
+                if !extra_found {
+                    extra_found = true;
+                }
+                state.add(nonmatching_output(*span, name, &context));
+            }
+        }
+
+        // Check for extra entries in `meta`.
+        for (name, span) in &actual {
+            if !expected.contains_key(name) {
+                if !extra_found {
+                    extra_found = true;
+                }
+                state.add(extra_output_in_meta(*span, name, &context));
+            }
+        }
+
+        // Check for out-of-order entries.
+        if !extra_found && !actual.keys().eq(expected.keys()) {
+            state.add(out_of_order(
+                meta_outputs_key.syntax().text_range().to_span(),
+                outputs.syntax().text_range().to_span(),
+                &context,
+            ));
+        }
+    } else {
+        state.add(missing_outputs_in_meta(
+            meta.syntax().first_token().unwrap().text_range().to_span(),
+            &context,
+        ));
+    }
+}
+
+impl Visitor for NonmatchingOutputRule {
+    type State = Diagnostics;
+
+    fn document(&mut self, _: &mut Self::State, reason: VisitReason, _: &Document) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        // Reset the visitor upon document entry
+        *self = Default::default();
+    }
+
+    fn workflow_definition(
+        &mut self,
+        state: &mut Self::State,
+        reason: VisitReason,
+        workflow: &WorkflowDefinition,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        // If the `meta` section is missing, the MissingMeta rule will handle it.
+        if let Some(meta) = workflow.metadata().next() {
+            // If the `output` section is missing, the MissingOutput rule will handle it.
+            if let Some(output) = workflow.outputs().next() {
```

Comment:
```suggestion
            if let Some(outputs) = workflow.outputs().next() {
```

to match the variable name in `check_output_meta`.

Reply:
I'm torn. This is an `OutputSection`, but is accessed with the `outputs()` method. I'm inclined to make the opposite update and make it consistently `output` when referring to the `OutputSection`.

Reply:
Throwing my 2 pennies in the ring: I like universally saying `output` (no 's')